scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala

```diff
@@ -1982,32 +2051,135 @@ class FormatOps(
         }
         if (ok) Some(new OptionalBraces(tree, true)) else None
       }
-      ftMeta.leftOwner match {
-        case t: Term.Match => result(t, t.cases)
-        case t: Type.Match => result(t, t.cases)
-        case _ => None
+      getMissingBrace(ftMeta).flatMap { _ =>
+        ftMeta.leftOwner match {
+          case t: Term.Match => result(t, t.cases)
+          case t: Type.Match => result(t, t.cases)
+          case _ => None
+        }
       }
     }
   }
 
+  // Optional braces after any token that can start indentation:
+  // then
+  object OptionalBracesDo {
+    def unapply(
+        ftMeta: FormatToken.Meta
+    )(implicit style: ScalafmtConfig): Option[OptionalBraces] =
+      getMissingBrace(ftMeta).flatMap { nft =>
+        (ftMeta.leftOwner match {
+          case t: Term.Do => Some(t.body -> true)
+          case t: Term.While => Some(t.body -> false)
+          case t: Term.For => Some(t.body -> false)
+          case _ => None
+        }).map { case (body, allowMain) =>
+          val singleStat = !isTreeMultiStatBlock(body)
+          val forceNL = !singleStat || shouldBreakInOptionalBraces(nft)
+          new OptionalBraces(body, forceNL, allowMain && singleStat)
+        }
+      }
+  }
+
+  // Optional braces after any token that can start indentation:
+  // then
+  object OptionalBracesThen {
+    def unapply(
+        ftMeta: FormatToken.Meta
+    )(implicit style: ScalafmtConfig): Option[OptionalBraces] =
+      getMissingBrace(ftMeta).flatMap { nft =>
+        ftMeta.leftOwner match {
+          case t: Term.If =>
+            def nestedIf(x: Term.If) = {
+              val forceNL = shouldBreakInOptionalBraces(nft) ||
+                !ifWithoutElse(t) && existsIfWithoutElse(x)
+              Some(new OptionalBraces(x, forceNL))
+            }
+            t.thenp match {
+              case x: Term.If => nestedIf(x)
+              case Term.Block(List(x: Term.If)) => nestedIf(x)
```

Comment:
Also, what was the issue with nested ifs? It seems it should work normally :thinking:

Reply:
this covers a block which contains only an if, not one which simply starts with an if.

if the block has multiple statements, a break is a must.

if the block only has an if, you must break, too, if it doesn't have an else but the outer one does:
```
if a then if b then c
else d
```
this `else` could belong to either of them since their indent is the same.